Add archiving capabilities

retogerber opened the pull request:
## Ticket

Will solve #26 

## Description

Idea is to retrieve files and store as single zip archive. Adds the following cli: 
```
Usage: ob info archive [OPTIONS]

  Archive a benchmark

Options:
  -b, --benchmark PATH            Path to benchmark yaml file or
                                  benchmark id.  [required]
  -c, --code                      Archive benchmarking code (repos).
  -s, --software                  Archive software environments.
  -r, --results                   Archive results files.
  --compression [none|deflated|bzip2|lzma]
                                  Compression method.  [default: none]
  --compresslevel INTEGER         Compression level.
  -n, --dry-run                   Do not create the archive, just show
                                  what would be done.
  --help                          Show this message and exit.
```

Also:
- moves helper functions from omni/workflow/scripts/run_module.py to omni/io/code.py
- allows to retrieve archives from github/gitlab instead of git cloning repos. (omni/io/code.py)


## Checklist:
- [x] My code follows the code style of this project.
- [x] My CLI method respects the signature defined in the task.
- [x] I have documented the CLI method accordingly.
- [x] I have added tests to cover my changes.
- [ ] I have added a CHANGELOG entry.

## Remarks for the reviewer:

Regarding software archiving, it just collects the files specified in the benchmark.yaml file and adds those to the archive.

DanInci commented:
Regarding software archiving, I wonder if it would be enough to only archive recipes. Meaning:
- In the case of conda, there's the conda yaml file, and its somewhat a recipe for building the conda environment.
- In the case of easybuild/envmodules, it would make sense to only archive the easybuild file (if available) since in theory its used to build the envmodules.
- In the case of docker/apptainer, this would mean having access to the Dockerfile used for building the image, but now thinking about it, we don't support that directly, as we do with easybuild+envmodules. Would it make sense to support that?

imallona commented on omni/cli/benchmark.py:
```diff
@@ -15,17 +16,49 @@ def info(ctx):
+    help="Path to benchmark yaml file or benchmark id.",
+    envvar="OB_BENCHMARK",
+)
+@click.option(
+    "-c",
+    "--code",
+    help="Archive code files.",
+    is_flag=True,
+    default=False,
+)
+@click.option(
+    "-s",
+    "--software",
+    help="Archive software files.",
+    is_flag=True,
+    default=False,
+)
```
Agree, I'd use "benchmarking code" and "software environments" instead

retogerber replied:
Right, currently it's not very descriptive, will make it more verbose.

retogerber commented:
> Regarding software archiving, I wonder if it would be enough to only archive recipes. Meaning:
> 
> * In the case of conda, there's the conda yaml file, and its somewhat a recipe for building the conda environment.
> * In the case of easybuild/envmodules, it would make sense to only archive the easybuild file (if available) since in theory its used to build the envmodules.

Thanks, yes I agree archiving the conda yaml and/or easybuild file should be enough.

> > * In the case of docker/apptainer, this would mean having access to the Dockerfile used for building the image, but now thinking about it, we don't support that directly, as we do with easybuild+envmodules. Would it make sense to support that?
> 
> We have an [`ob software singularity push`](https://omnibenchmark.org/reference/#push) already
Thanks, but I assume for archiving we would want to have everything in one place so I don't think pushing to some registry is what we want here. Since the Dockerfile is not necessarily enough to recreate an image I thought about directly storing the .sif image in the archive, what do you think @imallona @DanInci ?

imallona commented:
Thank you @retogerber ! Superuseful function to have! I have a couple of questions:
- why `ob info archive` - as an info tool? wouldn't be more user-friendly if named `ob storage archive` instead?
- could `ob info archive --compression` offer `gzip` as well?
- how do we know how to pick `ob info archive --compresslevel INTEGER`? is there a default?
- does it make sense this to be run locally? Because it does not offer the `--local` flag `ob run benchmark` has. More precisely, I was trying
```
(omnibenchmark) me@archiving:~/src/clustering_example$ ob info archive -b Clustering_conda.yml -c -s -r --compression bzip2 -n 
Invalid S3 config. Missing access_key and secret_key in environment variables (OB_STORAGE_S3_ACCESS_KEY, OB_STORAGE_S3_SECRET_KEY) or OB_STORAGE_S3_CONFIG

(omnibenchmark) me@archiving:~/src/clustering_example$ ob info archive -b Clustering_conda.yml -c -s -r --compression bzip2 -n --local
Usage: ob info archive [OPTIONS]
Try 'ob info archive --help' for help.

Error: No such option: --local
```

I'm trying to get the credentials to test this, will update.

retogerber commented:
Thanks for testing and the feedback! 
> * why `ob info archive` - as an info tool? wouldn't be more user-friendly if named `ob storage archive` instead?

agreed, moved to 'storage'
> * could `ob info archive --compression` offer `gzip` as well?

currently I just added all options available [here](https://docs.python.org/3/library/zipfile.html#zipfile.ZIP_STORED), but could also add gzip with [this](https://docs.python.org/3/library/gzip.html).
> * how do we know how to pick `ob info archive --compresslevel INTEGER`? is there a default?

The default should be the same as in `zipfile` which I think is 3.
> * does it make sense this to be run locally? Because it does not offer the `--local` flag `ob run benchmark` has. More precisely, I was trying

Good point, added.

imallona commented:
Also, I've noticed the archive files end with `zip` regarding the compression algo; worth changing the file extension to something compression-specific? e.g. `bz2` if `bzip2` etc?

imallona reviewed (approved):
Looks good to me!

Extra issue/comment/feedback, would it be possible to automate archiving the parameters.txt and parameters_dict.txt files? I'm afraid parsing results is hard without those, even if the yaml contains the param translation. What do you think?

retogerber commented:
> Looks good to me!
> 
> Extra issue/comment/feedback, would it be possible to automate archiving the parameters.txt and parameters_dict.txt files? I'm afraid parsing results is hard without those, even if the yaml contains the param translation. What do you think?

Thanks @imallona, how is the convention of the location of those files? For example for the test benchmark can we use the following:
```
>>> benchmark.get_nodes()[0].get_outputs()
['out/{stage}/{module}/{params}/{dataset}.txt.gz', 'out/{stage}/{module}/{params}/{dataset}.meta.json', 'out/{stage}/{module}/{params}/{dataset}_params.txt']
```
i.e. the parameters.txt would be located at 
```
'out/{stage}/{module}/{params}/parameters.txt'
```
and the dict at 
```
'out/{stage}/{module}/parameters_dict.txt
```
is this correct?

imallona commented:
yes it is, thanks @retogerber!
